r/R/dplyr-funcs-type.R

```diff
@@ -0,0 +1,252 @@
+    if (!is.null(check.rows)) arrow_not_supported("check.rows")
+
+    args <- rlang::dots_list(..., .named = fix.empty.names)
+    if (is.null(names(args))) {
+      names(args) <- rep("", length(args))
+    }
+
+    if (identical(check.names, TRUE)) {
+      if (identical(fix.empty.names, TRUE)) {
+        names(args) <- make.names(names(args), unique = TRUE)
+      } else {
+        name_emtpy <- names(args) == ""
+        names(args)[!name_emtpy] <- make.names(names(args)[!name_emtpy], unique = TRUE)
+      }
+    }
+
+    build_expr(
+      "make_struct",
+      args = unname(args),
+      options = list(field_names = names(args))
+    )
+  })
+}
+
+register_bindings_type_inspect <- function() {
+  # is.* type functions
+  register_binding("is.character", function(x) {
+    is.character(x) || (inherits(x, "Expression") &&
+                          x$type_id() %in% Type[c("STRING", "LARGE_STRING")])
+  })
+  register_binding("is.numeric", function(x) {
+    is.numeric(x) || (inherits(x, "Expression") && x$type_id() %in% Type[c(
+      "UINT8", "INT8", "UINT16", "INT16", "UINT32", "INT32",
+      "UINT64", "INT64", "HALF_FLOAT", "FLOAT", "DOUBLE",
+      "DECIMAL128", "DECIMAL256"
+    )])
+  })
+  register_binding("is.double", function(x) {
+    is.double(x) || (inherits(x, "Expression") && x$type_id() == Type["DOUBLE"])
+  })
+  register_binding("is.integer", function(x) {
+    is.integer(x) || (inherits(x, "Expression") && x$type_id() %in% Type[c(
+      "UINT8", "INT8", "UINT16", "INT16", "UINT32", "INT32",
+      "UINT64", "INT64"
+    )])
+  })
+  register_binding("is.integer64", function(x) {
+    inherits(x, "integer64") || (inherits(x, "Expression") && x$type_id() == Type["INT64"])
+  })
+  register_binding("is.logical", function(x) {
+    is.logical(x) || (inherits(x, "Expression") && x$type_id() == Type["BOOL"])
+  })
+  register_binding("is.factor", function(x) {
+    is.factor(x) || (inherits(x, "Expression") && x$type_id() == Type["DICTIONARY"])
+  })
+  register_binding("is.list", function(x) {
+    is.list(x) || (inherits(x, "Expression") && x$type_id() %in% Type[c(
+      "LIST", "FIXED_SIZE_LIST", "LARGE_LIST"
+    )])
+  })
+
+  # rlang::is_* type functions
+  register_binding("is_character", function(x, n = NULL) {
+    assert_that(is.null(n))
+    call_binding("is.character", x)
+  })
+  register_binding("is_double", function(x, n = NULL, finite = NULL) {
+    assert_that(is.null(n) && is.null(finite))
+    call_binding("is.double", x)
+  })
+  register_binding("is_integer", function(x, n = NULL) {
+    assert_that(is.null(n))
+    call_binding("is.integer", x)
+  })
+  register_binding("is_list", function(x, n = NULL) {
+    assert_that(is.null(n))
+    call_binding("is.list", x)
+  })
+  register_binding("is_logical", function(x, n = NULL) {
+    assert_that(is.null(n))
+    call_binding("is.logical", x)
+  })
+}
+
+register_bindings_type_elementwise <- function() {
+
```

Comment:
```suggestion
```

Super minor, and we can keep the new line if you think it adds to these, but it looks superfluous to me